From 9ea1a675f40f55c4eade5c18f5f7e73253cabf13 Mon Sep 17 00:00:00 2001
From: William Bellavance <flintman@flintmancomputers.com>
Date: Tue, 30 Aug 2016 08:04:38 -0400
Subject: [PATCH] init: don't skip starting a service with no domain if
 permissive

Requires external/libselinux change as well

[Adrian DC] Preserve the log while permissive

Change-Id: I3f2887930e15d09014c2594141ba4acbbc8d6d9d
---
 init/service.cpp | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/init/service.cpp b/init/service.cpp
index ccc37b70cb4..ad9a3e0aff2 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -94,11 +94,16 @@ static Result<std::string> ComputeContextFromExecutable(const std::string& servi
         free(new_con);
     }
     if (rc == 0 && computed_context == mycon.get()) {
#if ALLOW_PERMISSIVE_SELINUX
	// Allow permissive don't return error
#else
-        return Error() << "File " << service_path << "(labeled \"" << filecon.get()
-                       << "\") has incorrect label or no domain transition from " << mycon.get()
-                       << " to another SELinux domain defined. Have you configured your "
-                          "service correctly? https://source.android.com/security/selinux/"
-                          "device-policy#label_new_services_and_address_denials";
+        std::string error = StringPrintf(
+                "File %s (labeled \"%s\") has incorrect label or no domain transition from %s to "
+                "another SELinux domain defined. Have you configured your "
+                "service correctly? https://source.android.com/security/selinux/"
+                "device-policy#label_new_services_and_address_denials",
+                service_path.c_str(), filecon.get(), mycon.get());
+        if (selinux_status_getenforce() > 0) {
+            return Error() << error;
+        }
+        LOG(ERROR) << error;
#endif
    }
     if (rc < 0) {
         return Error() << "Could not get process context";
